### ISSUE TITLE
Usages of allOf within message payload could be flattened

### ISSUE THREAD

**Issue by jamescrowley** [asyncapi/asyncapi-react#596]
#### Describe the bug

I'm not sure if this is a bug or not, but for me, allOf is rendered in an unexpected way when used in the message payload. Can you confirm if this is deliberate or not? If not, I'll look to fix.

#### How to Reproduce

```
asyncapi: 2.0.0
info:
  title: MQTT API
  version: '2.0.0'

defaultContentType: application/json

channels:
  /v2/abc/data/pack_data:
    description: Publish logging data from devices
    publish:
      operationId: publishPackData
      message:
        name: pack_data
        payload:
          $ref: "#/components/schemas/pack_data"

components:
  schemas:
    pack_data:
      allOf:
      - $ref: "#/components/schemas/common_payload"
      - type: object
        properties:
          cycle_id:
            type: string
    common_payload:
      type: object
      properties:
        device_id:
            type: string
        site_id:
            type: string
```

<img width="789" alt="Screenshot 2020-04-23 at 13 47 24" src="https://user-images.githubusercontent.com/509533/80057011-f8791800-8568-11ea-925b-f341334cc078.png">

#### Expected behavior

I'd expect the use of 'allOf' to be transparent in the documentation (ie as if everything had just been defined explicitly in-line).

ie 
<img width="545" alt="Screenshot 2020-04-23 at 13 50 28" src="https://user-images.githubusercontent.com/509533/80057242-6ae9f800-8569-11ea-8ed6-4feb81af57e8.png">

**Comment by magicmatatjahu** [comment]
Still valid.

**Comment by fmvilas** [comment]
Still relevant.

**Comment by ChrisEdwards** [comment]
Why does this feature in AsyncAPI behave differently than it does in OpenAPI? Those two specs should behave the same. Users expect this, and it only causes confusion when they work differently.

**Comment by fmvilas** [comment]
They don't work differently. It's the same but the way **this component** handles it is different from the way other existing components for OpenAPI handle it. In all fairness, the way it's shown here is the correct way, even though I don't like it. This is not OpenAPI or AsyncAPI stuff, it's JSON Schema, and this is how JSON Schema defines it. JSON Schema is not a model or type definition language. It's a set of constraints and validations and, therefore, `allOf` can't be understood as merging all the objects. Again, I don't like it, but then we should probably escalate this to JSON Schema itself instead of this or other components.

@jonaslagoni started this initiative to clarify this on the JSON Schema side: https://github.com/json-schema-org/vocab-idl. It didn't have much traction so far so if you're really impacted by this issue it would be awesome if you could contribute there. If we get an IDL vocabulary, we'll for sure change AsyncAPI to use it and, as a result, this component would be able to merge all the `allOf` items.

**Comment by fmvilas** [comment]
In the meantime, we could add a flag to this repo to let the user decide how they want to treat `allOf`. Seems like a plausible medium-term solution.

**Comment by fmvilas** [comment]
Dear bot, close it not, por favor.

**Comment by henhal** [comment]
Any progress on this?
I just started creating specs for all my APIs today with AsyncAPI, and my initial excitement unfortunately faded quickly when I realized that the resulting HTML was pretty much unreadable. I structure my schemas heavily, relying on inheritance of base schemas etc, and to have documentation that does not merge allOfs in such a way that e.g. Redoc does it unfortunately makes it unusable to me.

I see that the PR has been open for several years. Is there a plan forward? Is there any debate of whether HTML should merge allOfs into the most restrictive combined schema? 🤔 For example, the highly used https://www.npmjs.com/package/json-schema-merge-allof handles the logic for allOf merging.

```
const mergeAllof = require('json-schema-merge-allof');

mergeAllOf({
  allOf: [
    {type: 'object', properties: {foo: {type: 'string'}}}, 
    {type: 'object', properties: {bar: {type: 'string'}, foo: {enum: ['a', 'b']}}}
  ]
});
```
Output:
```
{
  "type":"object",
  "properties":{
    "foo":{"type":"string","enum":["a","b"]},
    "bar":{"type":"string"}
  }
}
```
